sklearn/ensemble/_gradient_boosting.pyx

```diff
@@ -84,22 +88,99 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
     """
     cdef Py_ssize_t i
-    cdef int32 node_id
     cdef Node *node
     for i in range(n_samples):
         node = root_node
         # While node not a leaf
-        while node.left_child != -1 and node.right_child != -1:
+        while node.left_child != TREE_LEAF:
             if X[i * n_features + node.feature] <= node.threshold:
                 node = root_node + node.left_child
             else:
                 node = root_node + node.right_child
         out[i * K + k] += scale * value[node - root_node]
 
 
-@cython.nonecheck(False)
+def _predict_regression_tree_stages_sparse(np.ndarray[object, ndim=2] estimators,
+                                           object X, double scale,
+                                           np.ndarray[float64, ndim=2] out):
+    """Predicts output for regression tree inplace and adds scaled value to ``out[i, k]``.
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef np.ndarray[ndim=1, dtype=DTYPE_t] X_data_ndarray = X.data
+    cdef np.ndarray[ndim=1, dtype=INT32_t] X_indices_ndarray  = X.indices
+    cdef np.ndarray[ndim=1, dtype=INT32_t] X_indptr_ndarray  = X.indptr
+
+    cdef DTYPE_t* X_data = <DTYPE_t*>X_data_ndarray.data
+    cdef INT32_t* X_indices = <INT32_t*>X_indices_ndarray.data
+    cdef INT32_t* X_indptr = <INT32_t*>X_indptr_ndarray.data
+
+    cdef SIZE_t n_samples = X.shape[0]
+    cdef SIZE_t n_features = X.shape[1]
+    cdef SIZE_t n_stages = estimators.shape[0]
+    cdef SIZE_t n_outputs = estimators.shape[1]
+
+    # Initialize output
+    cdef float64* out_ptr = <float64*> out.data
+
+    # Indices and temporary variables
+    cdef SIZE_t sample_idx = 0
+    cdef SIZE_t feature_idx = 0
+    cdef SIZE_t stage_idx = 0
+    cdef SIZE_t output_idx = 0
+    cdef Node *root_node = NULL
+    cdef Node *node = NULL
+    cdef double *value = NULL
+
+    # Initialize auxiliary data-structure
+    cdef DTYPE_t feature_value = 0.
+    cdef DTYPE_t* X_sample = NULL
+
+    # feature_to_sample as a data structure records the last seen sample
+    # for each feature; functionally, it is an efficient way to identify
+    # which features are nonzero in the present sample.
+    cdef SIZE_t* feature_to_sample = NULL
+
+    safe_realloc(&X_sample, n_features * sizeof(DTYPE_t))
+    safe_realloc(&feature_to_sample, n_features * sizeof(SIZE_t))
+
+    memset(feature_to_sample, -1, n_features * sizeof(SIZE_t))
+
+    # Cycle through all samples
+    for sample_idx in range(n_samples):
+        for feature_idx in range(X_indptr[sample_idx], X_indptr[sample_idx + 1]):
+            feature_to_sample[X_indices[feature_idx]] = sample_idx
+            X_sample[X_indices[feature_idx]] = X_data[feature_idx]
+
+        # Cycle through all stages
+        for stage_idx in range(n_stages):
+            # Cycle through all trees
+            for output_idx in range(n_outputs):
+                root_node = (<Tree> estimators[stage_idx, output_idx].tree_).nodes
```

Comment:
I note that this is native Python work, but probably negligible. If we find  slowness, we might consider materialising an array of `Tree`s outside the loop over samples.

Also: pep8 line length